go/oasis-node/cmd/debug/byzantine/storage_node.go

```diff
@@ -0,0 +1,143 @@
+package byzantine
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path/filepath"
+	"sync"
+
+	flag "github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	"github.com/oasisprotocol/oasis-core/go/common/identity"
+	"github.com/oasisprotocol/oasis-core/go/storage/api"
+	storage "github.com/oasisprotocol/oasis-core/go/storage/api"
+	"github.com/oasisprotocol/oasis-core/go/storage/database"
+	"github.com/oasisprotocol/oasis-core/go/storage/mkvs/checkpoint"
+	"github.com/oasisprotocol/oasis-core/go/storage/mkvs/syncer"
+)
+
+const (
+	// CfgNumStorageFailApplyBatch configures how many apply-batch requests
+	// the storage node should fail.
+	CfgNumStorageFailApplyBatch = "num_storage_fail_apply_batch"
+	// CfgFailReadRequests configures if storage node should fail read requests.
+	CfgFailReadRequests = "fail_read_requests"
+)
+
+var (
+	_ storage.Backend = (*storageWorker)(nil)
+
+	errByzantine = fmt.Errorf("byzantine error")
+
+	storageFlags = flag.NewFlagSet("", flag.ContinueOnError)
+)
+
+type storageWorker struct {
+	sync.Mutex
+
+	id      *identity.Identity
+	backend storage.Backend
+	initCh  chan struct{}
+
+	numFailApplyBatch uint64
+	failReadRequests  bool
+}
+
+func newStorageNode(id *identity.Identity, namespace common.Namespace, datadir string) (*storageWorker, error) {
+	initCh := make(chan struct{})
+	defer close(initCh)
+
+	cfg := &api.Config{
+		Backend:           database.BackendNameBadgerDB,
+		DB:                filepath.Join(datadir, database.DefaultFileName(database.BackendNameBadgerDB)),
+		Signer:            id.NodeSigner,
+		ApplyLockLRUSlots: uint64(1000),
+		Namespace:         namespace,
+		MaxCacheSize:      64 * 1024 * 1024,
+	}
+	impl, err := database.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageWorker{
+		id:                id,
+		backend:           impl,
+		initCh:            initCh,
+		numFailApplyBatch: viper.GetUint64(CfgNumStorageFailApplyBatch),
+	}, nil
+}
+
+func (w *storageWorker) SyncGet(ctx context.Context, request *syncer.GetRequest) (*syncer.ProofResponse, error) {
+	if w.failReadRequests {
+		return nil, errByzantine
+	}
+
+	return w.backend.SyncGet(ctx, request)
+}
+
+func (w *storageWorker) SyncGetPrefixes(ctx context.Context, request *syncer.GetPrefixesRequest) (*syncer.ProofResponse, error) {
+	if w.failReadRequests {
+		return nil, errByzantine
+	}
+
+	return w.backend.SyncGetPrefixes(ctx, request)
+}
+
+func (w *storageWorker) SyncIterate(ctx context.Context, request *syncer.IterateRequest) (*syncer.ProofResponse, error) {
+	if w.failReadRequests {
+		return nil, errByzantine
+	}
+
+	return w.backend.SyncIterate(ctx, request)
+}
+
+func (w *storageWorker) Apply(ctx context.Context, request *storage.ApplyRequest) ([]*storage.Receipt, error) {
+	return w.backend.Apply(ctx, request)
```

Comment:
Should we also test with failing applies (this is used when creating the transaction batch)?

Reply:
Added it.

Didn't include it before as nothing interesting really happens, as the proposer should just retry proposing batches until Apply succeeds, and this is something that's regularly tested in the daily e2e tests. But nevertheless doesn't hurt adding it.